### PR TITLE
Add support for admission controller based on namespace node selectors.

### PR DESCRIPTION
**Pull request by aveshagarwal**
This work is to upstream openshift's project node selectors based admission controller.

Fixes https://github.com/kubernetes/kubernetes/issues/17151

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/24980)

### CODE REVIEW

**Comment by aveshagarwal** [comment]
@kubernetes/rh-cluster-infra

**Comment by aveshagarwal** [comment]
Added test cases for namespace cahce and labels.

File: pkg/labels/selector.go
Comment (DirectXMan12) [inline pkg/labels/selector.go]: You should note that the second map `wins` when conflicts do occur.

File: pkg/labels/selector.go
Comment (DirectXMan12) [inline pkg/labels/selector.go]: Nit: should be ConvertSelector**To**LabelsMap

File: pkg/labels/selector.go
Comment (DirectXMan12) [inline pkg/labels/selector.go]: Nit: Doc should be in the form of "ConvertSelectorToLabelsMap frobs the bits..." (start with the method name).

File: pkg/labels/selector.go
Comment (aveshagarwal) [inline pkg/labels/selector.go]: Here it is just to find out if a conflict occurs (true or false) that means if  a same key exists, then their values are same or not. Not sure what do you mean by win here?

Comment (aveshagarwal) [inline pkg/labels/selector.go]: @DirectXMan12 nv I was thinking about conflicts.

**Comment by aveshagarwal** [comment]
@DirectXMan12 thanks. Updated.

**Comment by aveshagarwal** [comment]
I noticed that selectors like "a=b, c=d, x=y" are valid as it does not have to be "a=b,c=d,x=y" always. So added white space trimming and now it brings it to parity with the openshift implementation w.r.t label processing. Also added more unit tests.

File: pkg/labels/selector.go
Comment (ncdc) [inline pkg/labels/selector.go]: When you view the godoc as rendered html in a web browser, it shows up like this:

"Conflicts takes 2 maps returns true if there a key match between the maps but the value doesn't match returns false in other cases"

It runs together. Please update accordingly.

Comment (ncdc) [inline pkg/labels/selector.go]: This also only checks the keys in the first map. If you have keys in the second map that don't exist in the first, this function doesn't consider that a conflict.

File: pkg/labels/selector.go
Comment (ncdc) [inline pkg/labels/selector.go]: Godoc needs formatting update

File: pkg/labels/selector.go
Comment (ncdc) [inline pkg/labels/selector.go]:
```go
if len(selector) == 0 {
  return labelsMap
}
...
```

File: plugin/pkg/admission/nodeenv/interfaces.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/interfaces.go]: NamespaceGetter

Comment (ncdc) [inline plugin/pkg/admission/nodeenv/interfaces.go]: Godoc

File: plugin/pkg/admission/nodeenv/interfaces.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/interfaces.go]: Godoc

File: plugin/pkg/admission/nodeenv/admission.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: Wait to do this at the end of the reviews, but I want to pick a different name for this package.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: Add a log message that you expected a pod but got a ...

File: plugin/pkg/admission/nodeenv/admission.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: do not export

File: plugin/pkg/admission/nodeenv/admission.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: do not export

File: plugin/pkg/admission/nodeenv/admission.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: I think it makes sense to model this after https://github.com/kubernetes/kubernetes/blob/0061479890ddf827bd55570ccbf8bd59352543c3/plugin/pkg/admission/namespace/lifecycle/admission.go#L114, i.e. just take in the client, and create a store/reflector in the body of the method.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: Create a store above and set it here, like this: https://github.com/kubernetes/kubernetes/blob/0061479890ddf827bd55570ccbf8bd59352543c3/plugin/pkg/admission/namespace/lifecycle/admission.go#L115-L129

File: plugin/pkg/admission/nodeenv/admission.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: This should be something the cluster admin can set when configuring the cluster. It needs to come from `config` on line 15 above

File: plugin/pkg/admission/nodeenv/admission_test.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission_test.go]: add a field for defaultNodeSelector

File: plugin/pkg/admission/nodeenv/admission_test.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission_test.go]: Isn't this case the same as line 48?

File: plugin/pkg/admission/nodeenv/admission_test.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission_test.go]: Isn't this case the same as line 48 (except instead of infra=false, this is infra=true).

File: plugin/pkg/admission/nodeenv/admission_test.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission_test.go]: I would omit this since admit=false - it's a bit confusing to see it here (and add a `continue` after line 111 if !test.admit)

File: pkg/labels/selector.go
Comment (aveshagarwal) [inline pkg/labels/selector.go]: yeah because that is not a conflict, only same keys cause conflict so iterating on one map is fine.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (aveshagarwal) [inline plugin/pkg/admission/nodeenv/admission.go]: I remember based on our previous discussion with @deads2k  that we do not want to create its own reflectors here. Not sure when this changed. Previous revision of this PR was doing it similarly only in a different file.  Anyway I have modified it and will send update soon.

Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: We'll do a reflector in a follow-up PR. It requires additional plumbing that we don't have right now in Kube.

**Comment by aveshagarwal** [comment]
Updated this PR.

**Comment by derekwaynecarr** [comment]
I will review this tomorrow.

**Comment by bgrant0607** [comment]
cc @davidopp

**Comment by derekwaynecarr** [comment]
I have yet to do a review on this, may not make top of my queue until tomorrow.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (derekwaynecarr) [inline plugin/pkg/admission/nodeenv/admission.go]: can you add a comment what this value is?

Comment (aveshagarwal) [inline plugin/pkg/admission/nodeenv/admission.go]: yeah sure.

Comment (aveshagarwal) [inline plugin/pkg/admission/nodeenv/admission.go]: Sorry, not sure you meant to add the comment in the code or here, 

Anyway defaultNodeSelector is the global default value (currently default to "") that is assigned to a name space if it does not have any at the time of its creation.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: Say what type the something else is

Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: Also this is an Errorf, not a Warning

File: plugin/pkg/admission/nodeenv/admission.go
Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: The default can't be hardcoded to "" above. It needs to come from the config.

Comment (derekwaynecarr) [inline plugin/pkg/admission/nodeenv/admission.go]: I meant in the code... I also was trying to work out why the default was ""

File: plugin/pkg/admission/nodeenv/admission.go
Comment (derekwaynecarr) [inline plugin/pkg/admission/nodeenv/admission.go]: can we get a // TODO here to make this a shared cache?

File: plugin/pkg/admission/nodeenv/admission.go
Comment (derekwaynecarr) [inline plugin/pkg/admission/nodeenv/admission.go]: yeah, so that was not obvious.  maybe change the name to 'clusterDefaultNodeSelector' and add a comment to that effect.

Comment (aveshagarwal) [inline plugin/pkg/admission/nodeenv/admission.go]: ok, if we want to get it from config, do we want to add a command line flag too to allow admins to configure the default?

File: plugin/pkg/admission/nodeenv/admission.go
Comment (derekwaynecarr) [inline plugin/pkg/admission/nodeenv/admission.go]: I think we need to have a priority list of annotations to pull from to support our downstream usage previously using openshift.io/node-selector.

Can the `podNodeEnvironment` get configured with the slice of annotations to look at?  That way when this is started in OpenShift, we can have a single admission controller that is configured to look at both "kubernetes.io/node-selector" and "openshift.io/node-selector"?

**Comment by derekwaynecarr** [comment]
cc @davidopp @bgrant0607 - this is basically upstreaming the feature demonstrated here: 

https://blog.openshift.com/deploying-applications-to-specific-nodes/

File: plugin/pkg/admission/nodeenv/admission.go
Comment (aveshagarwal) [inline plugin/pkg/admission/nodeenv/admission.go]: Yeah that sounds good to me.

**Comment by davidopp** [comment]
Is this use case for this "dedicated nodes"?

Taints/tolerations PR is almost finished (#24134) and there will be an admission controller that assigns tolerations based on annotation on namespace. So if this is for the dedicated node use case, we will already have that covered using that mechanism. Not that it's the end of the world to have two approaches doing the same thing, but the taints/toleration approach is more general so we will probably want to encourage people to use that long-term.

I'm OOO through Friday so will have extremely limited bandwidth to discuss this.

**Comment by aveshagarwal** [comment]
I have incorporated latest comments by @derekwaynecarr @ncdc, and updated this PR. PTAL.

**Comment by aveshagarwal** [comment]
Fixed tests.

File: pkg/labels/selector.go
Comment (sttts) [inline pkg/labels/selector.go]: same here:
```
if len(l) != 2 {
  return ...
}
```

File: plugin/pkg/admission/nodeenv/admission.go
Comment (sttts) [inline plugin/pkg/admission/nodeenv/admission.go]: Public variables are bad style.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (sttts) [inline plugin/pkg/admission/nodeenv/admission.go]: is it necessary to make `PodNodeEnvironmentPluginConfig` public? Type is private.

Comment (ncdc) [inline plugin/pkg/admission/nodeenv/admission.go]: For json/yaml decoding

File: plugin/pkg/admission/nodeenv/admission.go
Comment (sttts) [inline plugin/pkg/admission/nodeenv/admission.go]: Feels strange to return a string here. Why not make it `pluginConfig`? That would make the big godoc unnecessary above.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (sttts) [inline plugin/pkg/admission/nodeenv/admission.go]: blank line doesn't make sense here.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (deads2k) [inline plugin/pkg/admission/nodeenv/admission.go]: > Public variables are bad style.

This needs to be available as a plug point downstream for us to support our existing annotation.  Kube lacks the config concept like https://github.com/openshift/origin/blob/master/pkg/cmd/server/api/types.go#L1175, so actually using the controller config file is comparatively difficult.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (deads2k) [inline plugin/pkg/admission/nodeenv/admission.go]: > For json/yaml decoding

Unversioned without a scheme?

Comment (deads2k) [inline plugin/pkg/admission/nodeenv/admission.go]: > Unversioned without a scheme?

Ah, you're having to do this because kube still has a single file that's trying to be the config for every admission plugin so you can't have separate apiVersion/kind?

File: plugin/pkg/admission/nodeenv/admission.go
Comment (sttts) [inline plugin/pkg/admission/nodeenv/admission.go]: blank line unnecessary

File: plugin/pkg/admission/nodeenv/admission.go
Comment (sttts) [inline plugin/pkg/admission/nodeenv/admission.go]: How about following https://github.com/kubernetes/kubernetes/blob/4970f0c02d5384763b9833a35c639cea95df1e0f/plugin/pkg/admission/security/podsecuritypolicy/admission.go#L49 having `Run()` outside of the constructor. This looks like leaking go routines, plus without stop channel.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (aveshagarwal) [inline plugin/pkg/admission/nodeenv/admission.go]: separate apiVersion/kind just for configuring default cluster value for a plugin?

Comment (deads2k) [inline plugin/pkg/admission/nodeenv/admission.go]: > separate apiVersion/kind just for configuring default cluster value for a plugin?

Yes, different configs and objects are different.  Add to that the fact that you probably want a field for choosing annotations that has conditional defaulting and it starts looking a lot like an API object.

Comment (deads2k) [inline plugin/pkg/admission/nodeenv/admission.go]: > separate apiVersion/kind just for configuring default cluster value for a plugin?

Unfortunately, since kube works on a single shared controller config file, you have no reasonable way to manage that.

**Comment by aveshagarwal** [comment]
Updated based on feedback, rebased and fixed unit tests.
@derekwaynecarr @sttts @ncdc PTAL

**Comment by aveshagarwal** [comment]
Fixed failing tests and updated this PR.

**Comment by aveshagarwal** [comment]
@derekwaynecarr PTAL.

**Comment by derekwaynecarr** [comment]
@aveshagarwal - I am going to revisit this once kube 1.3 is cut since it wont make this milestone.

**Comment by aveshagarwal** [comment]
@derekwaynecarr rebased and retested PTAL.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (smarterclayton) [inline plugin/pkg/admission/nodeenv/admission.go]: The name for this plugin should be podplacementrestriction or namespacepodplacement or something.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (smarterclayton) [inline plugin/pkg/admission/nodeenv/admission.go]: This needs to be in a policy namespace (scoped).

**Comment by aveshagarwal** [comment]
@k8s-bot Please keep it open.

**Comment by aveshagarwal** [comment]
Rebased this PR, updated it with shared informer cache, implemented whitelist verification and added new units tests. @derekwaynecarr  PTAL.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (smarterclayton) [inline plugin/pkg/admission/nodeenv/admission.go]: Please change the package name to be more appropriate.  Env doesn't really make sense in this context.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (smarterclayton) [inline plugin/pkg/admission/nodeenv/admission.go]: Going to have to be `alpha.pod` or `pod.alpha`.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (smarterclayton) [inline plugin/pkg/admission/nodeenv/admission.go]: This should be a structured type, not a map[string].

File: plugin/pkg/admission/nodeenv/admission.go
Comment (aveshagarwal) [inline plugin/pkg/admission/nodeenv/admission.go]: Based on other annotation keys in pkg/api/helpers.go, shouldn't it be scheduler.alpha.kubernetes.io/node-selector like scheduler.alpha.kubernetes.io/tolerations and scheduler.alpha.kubernetes.io/affinity?

File: plugin/pkg/admission/nodeenv/admission.go
Comment (aveshagarwal) [inline plugin/pkg/admission/nodeenv/admission.go]: But this requires to be merged with pod.Spec.NodeSelector which is map[string]string?

**Comment by derekwaynecarr** [comment]
I'll play devils advocate, why alpha?  What would we really do to promote
this from alpha to beta to final form?  Do we really see the final form
being promoted directly to the NamespaceSpec?

**Comment by aveshagarwal** [comment]
@derekwaynecarr +1

**Comment by aveshagarwal** [comment]
@derekwaynecarr @smarterclayton so alpha or not alpha? I agree with Derek regarding why alpha  as we have not discussed any path forward yet to take it beyond alpha if we end up doing it alpha. Or it is a general practice to set it alpha as we are using annotations? Also any other feedback?

**Comment by aveshagarwal** [comment]
rebased.

**Comment by derekwaynecarr** [comment]
@aveshagarwal -- let's stick with alpha now.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (derekwaynecarr) [inline plugin/pkg/admission/nodeenv/admission.go]: can we move the package to `plugin/pkg/admission/namespace/nodeselector` ?

**Comment by aveshagarwal** [comment]
@derekwaynecarr so which one: scheduler.alpha.kubernetes.io/node-selector  or pod.alpha.kubernetes.io/node-selector ?

File: plugin/pkg/admission/nodeenv/admission.go
Comment (derekwaynecarr) [inline plugin/pkg/admission/nodeenv/admission.go]: or `plugin/pkg/admission/namespace/podnodeselector`

File: pkg/labels/selector.go
Comment (derekwaynecarr) [inline pkg/labels/selector.go]: should this be `labels.Set` ?

File: pkg/labels/selector.go
Comment (derekwaynecarr) [inline pkg/labels/selector.go]: same here, just make this `labels.Set`

Comment (derekwaynecarr) [inline pkg/labels/selector.go]: the same comment applies throughout.

File: plugin/pkg/admission/nodeenv/admission.go
Comment (derekwaynecarr) [inline plugin/pkg/admission/nodeenv/admission.go]: look at labels.Set?  i think that would be cleaner to use.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
few more updates then lgtm

File: plugin/pkg/admission/nodeenv/admission.go
Comment (aveshagarwal) [inline plugin/pkg/admission/nodeenv/admission.go]: going with podnodeselector .

**Comment by aveshagarwal** [comment]
@k8s-bot kubemark e2e test this

**Comment by aveshagarwal** [comment]
@derekwaynecarr updated it. PTAL.

**Comment by aveshagarwal** [comment]
The following errors dont seem related to a flake.
@k8s-bot kubemark e2e test this

**Comment by aveshagarwal** [comment]
@derekwaynecarr @smarterclayton Would appreciate if you could tell me if anything else is needed to get it in?

**Comment by davidopp** [comment]
Per our discussion at the SIG meeting today, this doesn't need an LGTM from me, feel free to merge once the appropriate RedHat people are happy with it.

**Comment by derekwaynecarr** [comment]
Reviewing this now.

**Review by derekwaynecarr — APPROVED** [review]

**Comment by derekwaynecarr** [comment]
LGTM

**Comment by aveshagarwal** [comment]
@k8s-bot test this issue: #IGNORE

**Comment by aveshagarwal** [comment]
the pr https://github.com/kubernetes/kubernetes/pull/32555 merged and changed the signature of admission.NewPluginInitializer which is now breaking the build of this admission controller. Will fix it soon.

**Comment by aveshagarwal** [comment]
@derekwaynecarr  this pr needs LGTM again.

**Pull request merged**
